guides/unit-testing.md

```diff
@@ -0,0 +1,973 @@
+# Unit Testing Guidelines
+
+## Use Jest
+
+[Jest](https://jestjs.io/) is the testing framework of choice for unit tests at MetaMask. It offers a more comprehensive set of features out of the box over Mocha or Tape that make writing tests for production code very easy. Some of the useful features are:
+
+- [Module mocks](https://jestjs.io/docs/next/mock-functions#mocking-modules)
+- [Timer mocks](https://jestjs.io/docs/next/timer-mocks)
+- [Snapshots](https://jestjs.io/docs/next/snapshot-testing)
+- Automatic [parallelization of tests](https://jestjs.io/blog/2016/09/01/jest-15#buffered-console-messages) without [restricting usage of exclusive tests](https://mochajs.org/#exclusive-tests-are-disallowed)
+- [Automatic sandboxing of tests](https://jestjs.io/blog/2016/03/11/javascript-unit-testing-performance) to prevent accidental leaks when modules are mocked
```

Comment:
Hmm... not sure about this one. I know Jest runs each test module in a "sandbox" but the individual tests within that module aren't isolated from each other automatically.

It's hard to tell what this linked post is referring to - maybe it's talking about how their virtualized environment allows effectively mocking `require` and resetting it between each test run? But even that doesn't happen automatically (AFAIK), you need to set `resetMocks` or manually call that between tests.

Reply:
Hmm, yes, that link isn't very useful. I might have meant something else.

In any case, you're right that only test files are isolated, not tests themselves. There is probably some reason why the Jest team decided to add sandboxing but it seems like an implementation detail. I can remove this.

Reply:
Removed here: [18feed3](https://github.com/MetaMask/contributor-docs/pull/38/commits/18feed3683747443f588f8d6a62a1908602507fe)